File: ai/tools.py
import subprocess
import shlex
import urllib.request
import urllib.parse
from config import load_config

# -------------------------------------------------
# Interactive commands blacklist
# -------------------------------------------------
INTERACTIVE_COMMANDS = {
    "vim", "vi", "nvim", "nano", "emacs",
    "top", "htop", "btop", "nvtop",
    "less", "more", "man",
    "ssh", "ftp", "telnet",
    "tmux", "screen",
    "watch", 
}

def _is_interactive(command: str) -> bool:
    try:
        argv = shlex.split(command)
    except ValueError:
        return True # Safe default
    
    if not argv:
        return False
        
    return argv[0] in INTERACTIVE_COMMANDS

# -------------------------------------------------
# Stable-mode allowlist (known, read-only commands)
# -------------------------------------------------

STABLE_ALLOWLIST = {
    # system info
    "fastfetch",
    "neofetch",
    "uname",
    "uptime",
    "lsb_release",
    "hostname",

    # memory / cpu
    "free",
    "vmstat",
    "mpstat",
    "lscpu",
    "lsmem",

    # disk
    "df",
    "lsblk",
    "mount",
    "findmnt",

    # processes (read-only)
    "ps",

    # networking (read-only)
    "ip",
    "ss",
    "iw",

    # files (non-destructive)
    "ls",
    "stat",
    "du",
    "tree",
    "cat",
    "head",
    "tail",
    "wc",
}

def _is_allowed(command: str) -> bool:
    """
    Stable:
      - Only known, read-only commands
    Experimental:
      - Anything goes EXCEPT interactive commands
    """
    cfg = load_config()
    
    if cfg.mode == "experimental":
        return not _is_interactive(command)

    try:
        argv = shlex.split(command)
    except ValueError:
        return False

    if not argv:
        return False

    base_cmd = argv[0]
    
    # Check config override first
    if cfg.allowed_commands:
        return base_cmd in cfg.allowed_commands
        
    return base_cmd in STABLE_ALLOWLIST


def run_shell_command(command: str) -> str:
    """
    Execute a shell command with mode-based safety policy.
    """
    if not command or not isinstance(command, str):
        return "[ERROR] Invalid command"

    if not _is_allowed(command):
        return (
            "[DENIED] Command not allowed in STABLE mode\n"
            "Enable experimental mode to allow unrestricted execution."
        )
    
    # Load config for timeout
    cfg = load_config()

    try:
        process = subprocess.run(
            command,
            shell=True,
            executable="/bin/bash",
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=cfg.timeout_seconds,
        )

        stdout = process.stdout.strip()
        stderr = process.stderr.strip()

        if stdout and stderr:
            return f"{stdout}\n\n[STDERR]\n{stderr}"
        if stdout:
            return stdout
        if stderr:
            return f"[STDERR]\n{stderr}"

        return "[INFO] Command executed with no output"

    except subprocess.TimeoutExpired:
        return "[ERROR] Command timed out"
    except Exception as e:
        return f"[ERROR] {type(e).__name__}: {str(e)}"

def search_web(query: str) -> str:
    """
    Basic web search tool (Mock/Placeholder or simple naive implementation).
    """
    # This is a placeholder. In a real scenario, you'd use a search API.
    # We can simulate it or just return a message if not configured.
    return f"[INFO] Searching the web for: {query}\n(Note: Real web search requires an API key configuration. This is a placeholder.)"

File: src/tools.rs
use std::process::{Command, Stdio};
use std::collections::HashSet;
use std::time::Duration;
use crate::config::AppConfig;
use anyhow::Result;

lazy_static::lazy_static! {
    static ref STABLE_ALLOWLIST: HashSet<&'static str> = {
        [
            "fastfetch", "neofetch", "uname", "uptime", "lsb_release", "hostname",
            "free", "vmstat", "mpstat", "lscpu", "lsmem",
            "df", "lsblk", "mount", "findmnt",
            "ps",
            "ip", "ss", "iw",
            "ls", "stat", "du", "tree", "cat", "head", "tail", "wc"
        ].into_iter().collect()
    };

    static ref INTERACTIVE_COMMANDS: HashSet<&'static str> = {
        [
            "vim", "vi", "nvim", "nano", "emacs",
            "top", "htop", "btop", "nvtop",
            "less", "more", "man",
            "ssh", "ftp", "telnet",
            "tmux", "screen",
            "watch"
        ].into_iter().collect()
    };
}

fn is_interactive(command: &str) -> bool {
    let parts: Vec<&str> = command.split_whitespace().collect();
    if parts.is_empty() {
        return false;
    }
    INTERACTIVE_COMMANDS.contains(parts[0])
}

fn is_allowed(command: &str, config: &AppConfig) -> bool {
    if config.mode == "experimental" {
        return !is_interactive(command);
    }

    let parts: Vec<&str> = command.split_whitespace().collect();
    if parts.is_empty() {
        return false;
    }
    let base_cmd = parts[0];

    if !config.allowed_commands.is_empty() {
        return config.allowed_commands.iter().any(|c| c == base_cmd);
    }

    STABLE_ALLOWLIST.contains(base_cmd)
}

pub async fn run_shell_command(command: &str, config: &AppConfig) -> String {
    if !is_allowed(command, config) {
        return format!(
            "[DENIED] Command not allowed in {} mode",
            config.mode
        );
    }

    // In a real async runtime, we might want to use tokio::process
    // keeping it simple with std::process for now, but wrapping in timeout
    
    // NOTE: std::process::Command blocks, but for short commands it's okay-ish.
    // For better async, use tokio::process::Command.
    let output_result = tokio::time::timeout(
        Duration::from_secs(config.timeout_seconds),
        async {
            tokio::process::Command::new("bash")
                .arg("-c")
                .arg(command)
                .stdout(Stdio::piped())
                .stderr(Stdio::piped())
                .output()
                .await
        }
    ).await;

    match output_result {
        Ok(Ok(output)) => {
            let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
            let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();

            if !stdout.is_empty() && !stderr.is_empty() {
                format!("{}\n\n[STDERR]\n{}", stdout, stderr)
            } else if !stdout.is_empty() {
                stdout
            } else if !stderr.is_empty() {
                format!("[STDERR]\n{}", stderr)
            } else {
                "[INFO] Command executed with no output".to_string()
            }
        }
        Ok(Err(e)) => format!("[ERROR] Execution failed: {}", e),
        Err(_) => "[ERROR] Command timed out".to_string(),
    }
}

pub fn search_web(query: &str) -> String {
    format!("[INFO] Searching the web for: {}\n(Placeholder)", query)
}
